src/awx_plugins/credentials/aws_assumerole.py

```diff
@@ -0,0 +1,128 @@
+                                 ],
+                      'metadata': [{'id': 'identifier',
+                                    'label': 'Identifier',
+                                    'type': 'string',
+                                    'help_text': _('The name of the key in the assumed AWS role to fetch [AccessKeyId | SecretAccessKey | SessionToken].'),
+                                    },
+                                   ],
+                      'required': ['role_arn'],
+                      }
+
+
+def aws_assumerole_getcreds(access_key, secret_key, role_arn, external_id):
+    if (access_key is None or len(access_key) == 0) and (
+            secret_key is None or len(secret_key) == 0
+    ):
+        # Connect using credentials in the EE
+        connection = boto3.client(service_name='sts')
+    else:
+        # Connect to AWS using provided credentials
+        connection = boto3.client(
+            service_name='sts',
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key,
+        )
+    try:
+        response = connection.assume_role(
+            RoleArn=role_arn,
+            RoleSessionName='AAP_AWS_Role_Session1',
+            ExternalId=external_id,
+        )
+    except ClientError as ce:
+        raise ValueError(f'Got a bad client response from AWS: {ce.msg}.')
+
+    credentials = response.get('Credentials', {})
+
+    return credentials
+
+
+def aws_assumerole_backend(**kwargs):
+    """This backend function actually contacts AWS to assume a given role for
+    the specified user."""
+    access_key = kwargs.get('access_key')
+    secret_key = kwargs.get('secret_key')
+    role_arn = kwargs.get('role_arn')
+    external_id = kwargs.get('external_id')
+    identifier = kwargs.get('identifier')
+
+    # Generate a unique SHA256 hash for combo of user access key and ARN
+    # This should allow two users requesting the same ARN role to have
+    # separate credentials, and should allow the same user to request
+    # multiple roles.
+    #
+    credential_key_hash = hashlib.sha256(
+        (str(access_key or '') + role_arn).encode('utf-8'),
+    )
+    credential_key = credential_key_hash.hexdigest()
+
+    credentials = _aws_cred_cache.get(credential_key, None)
+
+    # If there are no credentials for this user/ARN *or* the credentials
+    # we have in the cache have expired, then we need to contact AWS again.
+    #
+    if (credentials is None) or (
+        credentials['Expiration'] < datetime.datetime.now(
+            credentials['Expiration'].tzinfo,
+        )
+    ):
+
+        credentials = aws_assumerole_getcreds(
+            access_key, secret_key, role_arn, external_id,
+        )
+
+        _aws_cred_cache[credential_key] = credentials
+
+    credentials = _aws_cred_cache.get(credential_key, None)
+
+    if identifier in credentials:
+        return credentials[identifier]
+
+    raise ValueError(f'Could not find a value for {identifier}.')
```

Comment:
This could be a nice way to chain the root cause in tracebacks:
```suggestion
    try:
        return credentials[identifier]
    except KeyError as key_err:
        raise ValueError(f'Could not find a value for {identifier}.') from key_err
```

Reply:
interesting, have never seen `from key_err` before. TIL

Reply:
Yeah, it was introduced around Python 3.3, I think. It sets an explicit `__cause__` attribute, which is then used instead of the implicitly set `__context__`. Many people consider this the best practice because you're vocal about the causes of exceptions.
The traceback shows both exceptions, but the text is slightly different. If you use `from`, it'll say that the previous exception in the trace was the direct cause. You can also use `from None` to discard the exception from the previous context so that it's not attached to the reraised one.